src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp

```diff
@@ -15,95 +15,138 @@
 #include "mlir/Dialect/Tosa/IR/TosaOps.h"
 #include "mlir/IR/Attributes.h"
 #include "mlir/IR/BuiltinTypes.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
 #include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
 #include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/NewShapeHelper.hpp"
 #include "llvm/ADT/ArrayRef.h"
+#include <src/Dialect/Mlir/IndexExpr.hpp>
 
 using namespace mlir;
 
 namespace onnx_mlir {
 
 namespace {
 
-class ONNXMaxPoolSingleOutOpLoweringToTOSA
-    : public OpConversionPattern<ONNXMaxPoolSingleOutOp> {
+// This calculates the values that need to be added to the padding in order to
+// simulate the ceil mode
+llvm::SmallVector<int64_t> getCeilConstants(llvm::ArrayRef<int64_t> inputShape,
+    const NewONNXMaxPoolSingleOutOpShapeHelper &shapeHelper) {
+  // This avoids having more if statements when creating the padding const.
+  if (!shapeHelper.ceilMode)
+    return llvm::SmallVector<int64_t>{0, 0};
+
+  SmallVector<int64_t, 2> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
+
+  // Get stride and pad vectors.
+  SmallVector<int64_t, 2> stridesVec = shapeHelper.strides;
+  SmallVector<int64_t, 4> padsVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // Check if the idiv_check for the output dimentions in
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_max_pool2d has no rest. If
```

Comment:
Eek - I don't know what `has no rest` means. Is this a defined term?

Reply:
Oh that was a 1:1 german to english translation that is not possible. The correct term is remainder